Remove blanket TypeScript exception for client test files

aduth opened the pull request:
## Proposed Changes

* Removes blanket TypeScript exceptions for test files, effectively enforcing type-checking for new test files and files already passing
* Exempts files with existing failures, to limit the scope of these changes and with the expectation that these will be resolved in future pull requests
* Include `@testing-library/jest-dom` types
* Removes unnecessary exclusions
    * TypeScript ignores `node_modules` by default 
    * The `search-index.js` no longer exists
    * The Storybook pattern only matches a literal `.stories.tsx` file, not as a suffix

## Why are these changes being made?

* Better autocompletion and availability of type annotations for configured `types` libraries (e.g. `toBeInDocument` matcher)
* Avoid AI generation cycles trying to resolve editor-flagged inline TypeScript errors, which may also try to add `@testing-library/jest-dom` as an inline import to the file
* Benefit from the same reason we use type-checking anywhere else; namely, correct use of functions and components under test

See: p1745346000251659-slack-C02DQP0FP

## Testing Instructions

Verify type-checking passes

* `yarn typecheck`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Add testing-library/jest-dom types for matcher enhancements

aduth committed:
Remove unnecessary exclusions

1. node_modules excluded by default
2. search-index.js no longer exists
3. stories pattern only matches the literal ".stories.tsx" file, not as a suffix

aduth committed:
Remove blanket TypeScript exception for client test files

Exempt existing failures, but start checking new files and continue checking files already passing

aduth commented:
Noting that this could potentially cause issues with stale branches after merging, for open pull requests that add test files that might fail after these changes are merged. I can plan to share a message on Slack encouraging people to rebase stale branches for any pull requests including `.ts` or `.tsx` test files.

mirka reviewed (commented):
Agree that this is a step in the right direction!

One question I had about the approach is whether we should consider a `@ts-nocheck TODO: Fix the errors` comment on each of the existing violating files, rather than a central exclude list hidden in the tsconfig. Otherwise, any devs that happen to  touch the violating files will not realize that there's a cleanup task to be done, or even have a false sense of security that the `.ts` files are properly typechecked.

Aaand TIL this repo has a [`@typescript-eslint/ban-ts-comment`](https://typescript-eslint.io/rules/ban-ts-comment/) lint in place, which may be worth revisiting as well. (There are plenty of `eslint-disable-next-line`s for this already 💀)

> The Storybook pattern only matches a literal `.stories.tsx` file, not as a suffix

😂

aduth commented:
@mirka I think that's a good point! I agree it'd be easy to overlook these exceptions unless someone is making an intentional effort to handle them.

aduth committed:
Allow ts-disabling comments iwth description

Tolerate if developer makes conscious choice and is able to justify through description

aduth committed:
Add inline comment to individual files instead of exclude list

aduth committed:
Relocate ts-nocheck comment below jest-environment

jest-environment must come first

tyxla commented on client/my-sites/checkout/get-thank-you-page-url/test/get-thank-you-page-url.ts:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This will need to be moved down, below the jest-environment directive block.

tyxla commented on client/my-sites/plans-features-main/components/test/plan-notice.tsx:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This needs to be moved below the jest-environment code block

tyxla commented on client/components/theme-tier/theme-tier-badge/test/index.tsx:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This needs to be moved below the jest-environment block

tyxla commented on client/my-sites/plans/jetpack-plans/plan-upgrade/test/index.tsx:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This needs to be moved below the jest-environment block

tyxla commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/test/unified-plans-step.tsx:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This needs to be moved below the jest-environment block

tyxla commented on client/sites/plan/test/index.tsx:
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck - TODO: Fix TypeScript issues
```
This needs to be moved below the jest-environment block

tyxla reviewed (commented):
There are a few instances where we need to keep the `@jest-environment` directive at the top of the file.

aduth commented:
> There are a few instances where we need to keep the `@jest-environment` directive at the top of the file.

Good catches! My search / replace pattern used in 4713b64 didn't account for comments or single-line syntax of the directive. I'll make another pass at the remaining ones.
